### ISSUE TITLE
model methods ignore cascade

### ISSUE THREAD

**Issue by cjuega** [sequelize/sequelize#9926]
## What are you doing?
I want to destroy associated records in cascade when using polymorphic associations. However, it seems instance and model destroy methods behave different when handling cascade.

It's kind of confusing to have different behavior between instance and model methods. I would expect both to behave the same.

I would agree this is not an issue with regular associations because sequelize relies on database FK management. That ensures associated records will be deleted by the database directly. However, the behavior is quite different when we talk about polymorphic associations. In that case, FK constraints are disabled and hence, associated records remain there.

A similar issue was opened two years ago but it was closed without a proper explanation: https://github.com/sequelize/sequelize/issues/5439. I have commented there too, but it seems nobody saw it because it's closed.

Use case:

```js
let ModelA = sequelize.define('modelA', {
      ...
      type: {
         type: DataTypes.STRING,
         allowNull: false,
         validate: {...}
   }, {
      classMethods: {
         associate: (models) => {
            models.modelA.belongsTo(models.modelB, {
               constraints: false
            });
         }
      }
   }
);

let ModelB = sequelize.define('modelB', {
      ...
   }, {
      classMethods: {
         associate: (models) => {
            models.modelB.hasMany(models.modelA, {
               onDelete: 'cascade',
               hooks: true,
               scope: {
                  type: 'some filtering value'
               }
            });
         }
      }
   }
);
```

## What do you expect to happen?
Code below deletes any associated modelA too.
```js
return ModelB.findById(4)
   .then((instance) => {
     return instance.destroy();
   });
```

Code below doesn't delete associated modelA.
```js
return ModelB.destroy({where: {id: 4}});
```

I would expect the two statements above to do exactly the same.

## What is actually happening?
instance method destroys associations in cascade. However model method doesn't.

**Comment by sushantdhiman** [comment]
Have you tried passing `individualHooks` to `Model.destroy` ?

**Comment by cjuega** [comment]
Yes, I did. My unit tests actually include `individualHooks: true` but it still fails. Having a deeper look into sequelize code, I don't believe that's enough. It seems model's methods disable `options.cascade`. I also tried `individualHooks: true` and `cascade: true` to `Model.destroy`, but it didn't work either.

I noticed that `Model.destroy` relies on `QueryInterface.bulkDelete` whereas `instance.destroy` relies on `QueryInterface.delete`. I believe the issue is there because `QueryInterface.delete` does take care of cascading, but `QueryInterface.bulkDelete` doesn't.

The only workaround I found is to implement cascade manually using `beforeDestroy`/`afterDestroy` hook plus running everything inside a transaction. The latter is to ensure deletion and cascade actions occur atomically within the database.

However, I would expect cascade to work for model methods too as it isn't said otherwise.

**Comment by cjuega** [comment]
Also, I'm wondering if this also affects to `onUpdate: 'cascade'` as I noticed the same differences between `QueryInterface.update` and `QueryInterface.bulkUpdate`

**Comment by sushantdhiman** [comment]
What kind of solution are you looking for? Sequelize should fire delete for those polymorphic association which aren't enforced with foreign key, mimicking cascade?

**Comment by cjuega** [comment]
exactly. That's what is actually happening in `instance.delete` according to this code:

```js
  delete(instance, tableName, identifier, options) {
    const cascades = [];
    const sql = this.QueryGenerator.deleteQuery(tableName, identifier, {}, instance.constructor);

    options = _.clone(options) || {};

    // Check for a restrict field
    if (!!instance.constructor && !!instance.constructor.associations) {
      const keys = Object.keys(instance.constructor.associations);
      const length = keys.length;
      let association;

      for (let i = 0; i < length; i++) {
        association = instance.constructor.associations[keys[i]];
        if (association.options && association.options.onDelete &&
          association.options.onDelete.toLowerCase() === 'cascade' &&
          association.options.useHooks === true) {
          cascades.push(association.accessors.get);
        }
      }
    }

    return Promise.each(cascades, cascade => {
      return instance[cascade](options).then(instances => {
        // Check for hasOne relationship with non-existing associate ("has zero")
        if (!instances) {
          return Promise.resolve();
        }

        if (!Array.isArray(instances)) instances = [instances];

        return Promise.each(instances, instance => instance.destroy(options));
      });
    }).then(() => {
      options.instance = instance;
      return this.sequelize.query(sql, options);
    });
  }
```

The code above ensures associated records are destroyed by Sequelize mimicking cascade. I would expect model's method to do the same, so people can use instance or model methods indistinctly (based on which one suits better for each scenario).

Please, note there is an issue with non-polymorphic associations too. However it has less impact because the database engine will perform cascading in the end. My point is both methods take care of cascading within Sequelize, or none does. But it's confusing to have one method that takes care of it and another that doesn't.

**Comment by cjuega** [comment]
@sushantdhiman, do you expect any other comment from my side? Do you have any other comment? I see this as a bug and it would be great if it gets fixed.

**Comment by oleg-koval** [comment]
@cjuega Have actually the same problem, did you find any solution? thanks!

**Comment by cjuega** [comment]
@oleg-koval, the only workaround I can think of is what I described in a previous [comment](https://github.com/sequelize/sequelize/issues/9926#issuecomment-424312224). Basically you have to emulate cascading using sequelize's hooks. Then invoking model methods with `individualHooks: true` will do the trick. If you are worry about data consistency, then you must ensure everything runs inside a transaction (model method + hook functions).

**Comment by Thyago** [comment]
I am having the exact same problem. I tried setting the hasMany and belongsTo in both objects and all the possible sequence of attributes I found. It simple does not delete.
When I add a forced validation error in the Class of the objects that are supposed to be deleted, it is triggered. So I guess they are being updated instead of being deleted...

**Comment by JSilversun** [comment]
Hello, I have this problem too. Actually I wanted to do something like: `Model.destroy({where:{id:{in:['1','2']}}})`; and I expected it to delete a cascade deletion of all the previous models.

I examined the code and I saw two destroy methods in the `model.js file`:

```js
/**
   * Delete multiple instances, or set their deletedAt timestamp to the current time if `paranoid` is enabled.
   *
   * @param  {Object}       options                         destroy options
   * @param  {Object}       [options.where]                 Filter the destroy
   * @param  {boolean}      [options.hooks=true]            Run before / after bulk destroy hooks?
   * @param  {boolean}      [options.individualHooks=false] If set to true, destroy will SELECT all records matching the where parameter and will execute before / after destroy hooks on each row
   * @param  {number}       [options.limit]                 How many rows to delete
   * @param  {boolean}      [options.force=false]           Delete instead of setting deletedAt to current timestamp (only applicable if `paranoid` is enabled)
   * @param  {boolean}      [options.truncate=false]        If set to true, dialects that support it will use TRUNCATE instead of DELETE FROM. If a table is truncated the where and limit options are ignored
   * @param  {boolean}      [options.cascade=false]         Only used in conjunction with TRUNCATE. Truncates  all tables that have foreign-key references to the named table, or to any tables added to the group due to CASCADE.
   * @param  {boolean}      [options.restartIdentity=false] Only used in conjunction with TRUNCATE. Automatically restart sequences owned by columns of the truncated table.
   * @param  {Transaction}  [options.transaction] Transaction to run query under
   * @param  {Function}     [options.logging=false]         A function that gets executed while running the query to log the sql.
   * @param  {boolean}      [options.benchmark=false]       Pass query execution time in milliseconds as second argument to logging function (options.logging).
   *
   * @returns {Promise<number>} The number of destroyed rows
   */
  static destroy(options) {
    options = Utils.cloneDeep(options);

    // ...

    let instances;

    return Promise.try(() => {
      // Run before hook
      if (options.hooks) {
        return this.runHooks('beforeBulkDestroy', options);
      }
    }).then(() => {
      // ...
      }
    }).then(() => {
      // ...
      }
      // Please notice this line calls the bulkDelete method in QueryInterface, this is the main difference between this method and the single destroy method
      return this.QueryInterface.bulkDelete(this.getTableName(options), options.where, options, this);
    }).tap(() => {
      // ...
    }).tap(() => {
      // ...
    });
  }
```

Instead of the single record destroy method which uses the `delete` method of `QueryInterface`:

/**
   * Destroy the row corresponding to this instance. Depending on your setting for paranoid, the row will either be completely deleted, or have its deletedAt timestamp set to the current time.
   *
   * @param {Object}      [options={}] destroy options
   * @param {boolean}     [options.force=false] If set to true, paranoid models will actually be deleted
   * @param {Function}    [options.logging=false] A function that gets executed while running the query to log the sql.
   * @param {Transaction} [options.transaction] Transaction to run query under
   * @param {string}      [options.searchPath=DEFAULT] An optional parameter to specify the schema search_path (Postgres only)
   *
   * @returns {Promise}
   */
  destroy(options) {
    options = Object.assign({
      hooks: true,
      force: false
    }, options);

    return Promise.try(() => {
      // ...
    }).then(() => {
      // ...
        ).then(([results, rowsUpdated]) => {
          //...
      // This is the line that calls the `delete` method in the QueryInterface class
      return this.constructor.QueryInterface.delete(this, this.constructor.getTableName(options), where, Object.assign({ type: QueryTypes.DELETE, limit: null }, options));
    }).tap(() => {
      // ...
    });
  }

@cjuega Cited the code of the `delete` method in the `QueryInterface` class:

```js
delete(instance, tableName, identifier, options) {
    const cascades = [];
    const sql = this.QueryGenerator.deleteQuery(tableName, identifier, {}, instance.constructor);

    options = _.clone(options) || {};

    // Check for a restrict field
    if (!!instance.constructor && !!instance.constructor.associations) {
      const keys = Object.keys(instance.constructor.associations);
      const length = keys.length;
      let association;

      for (let i = 0; i < length; i++) {
        association = instance.constructor.associations[keys[i]];
        if (association.options && association.options.onDelete &&
          association.options.onDelete.toLowerCase() === 'cascade' &&
          association.options.useHooks === true) {
          cascades.push(association.accessors.get);
        }
      }
    }

    return Promise.each(cascades, cascade => {
      return instance[cascade](options).then(instances => {
        // Check for hasOne relationship with non-existing associate ("has zero")
        if (!instances) {
          return Promise.resolve();
        }

        if (!Array.isArray(instances)) instances = [instances];

        return Promise.each(instances, instance => instance.destroy(options));
      });
    }).then(() => {
      options.instance = instance;
      return this.sequelize.query(sql, options);
    });
  }
```

This is the code in charge to "simulate" onDelete cascade, on the other hand, the method `bulkDelete` used by the multiple records destroy method in `model.js` doesn't consider the `onDelete` cascade at all:

```js
bulkDelete(tableName, where, options, model) {
    options = Utils.cloneDeep(options);
    options = _.defaults(options, { limit: null });

    if (options.truncate === true) {
      return this.sequelize.query(
        this.QueryGenerator.truncateTableQuery(tableName, options),
        options
      );
    }

    if (typeof identifier === 'object') where = Utils.cloneDeep(where);

    return this.sequelize.query(
      this.QueryGenerator.deleteQuery(tableName, where, options, model),
      options
    );
  }

  select(model, tableName, options) {
    options = Utils.cloneDeep(options);
    options.type = QueryTypes.SELECT;
    options.model = model;

    return this.sequelize.query(
      this.QueryGenerator.selectQuery(tableName, options, model),
      options
    );
  }
``` 

I think this should be in the documentation at least, I spent like 4 hours figuring out why this didn't work, I thought it had something to do with the hooks but I don't think that's the case.

Basically, the only way to do an onDelete cascade of multiple records is to use `findAll` method to get all the records first and then do something like:

```js
for(const record of records){
    record.destroy();
}
```

**Comment by papb** [comment]
Can someone please test if this is still happening in the newest 5.x version?

**Comment by cjuega** [comment]
@papb, I haven't tested it on v5 yet. I did check the code though. I haven't seen any changes in [model.js#destroy](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L2729) nor [query-interface.js#bulkDelete](https://github.com/sequelize/sequelize/blob/master/lib/query-interface.js#L1096) that says otherwise. So I expect this issue to be reproducible on sequelize v5 too.

**Comment by cjuega** [comment]
oh, wait! I actually did check it on `"sequelize": "^5.8.8",`. If I disable individualHooks (which are workarrounding the issue), my specs still fail. So yes, this is still happening.
